metals/src/main/scala/scala/meta/internal/metals/SemanticdbIndexer.scala

```diff
@@ -0,0 +1,67 @@
+package scala.meta.internal.metals
+
+import java.nio.file.Path
+import MetalsEnrichments._
+import java.nio.file.Files
+import ch.epfl.scala.bsp4j.ScalacOptionsResult
+import scala.meta.internal.semanticdb.TextDocuments
+
+class SemanticdbIndexer(
+    referenceProvider: ReferenceProvider,
+    implementationProvider: ImplementationProvider
+) {
+
+  def onScalacOptions(scalacOptions: ScalacOptionsResult): Unit = {
+    for {
+      item <- scalacOptions.getItems.asScala
+    } {
+      val targetroot = item.targetroot
+      onChangeDirectory(targetroot.resolve(Directories.semanticdb).toNIO)
+    }
+  }
+
+  def reset(): Unit = {
+    referenceProvider.reset()
+    implementationProvider.clear()
+  }
+
+  def onDelete(file: Path): Unit = {
+    referenceProvider.onDelete(file)
+    implementationProvider.onDelete(file)
+  }
+
+  /**
+   * Handle EventType.OVERFLOW, meaning we lost file events for a given path.
+   *
+   * We walk up the file tree to the parent `META-INF/semanticdb` parent directory
+   * and re-index all of its `*.semanticdb` children.
+   */
+  def onOverflow(path: Path): Unit = {
+    path.semanticdbRoot match {
+      case Some(root) =>
+        onChangeDirectory(root)
+      case None =>
+    }
+  }
+
+  def onChangeDirectory(dir: Path): Unit = {
+    if (Files.isDirectory(dir)) {
+      val stream = Files.walk(dir)
+      try {
+        stream.forEach(path => if (path.toFile().isFile()) onChange(path))
```

Comment:
we could hide the `isFile` check inside `onChange`:
```scala
def onChange(path: Path) = {
  if(!Files.isFile(path)) ()
  else (...)
}
```

or at least we could use `Files.isFile(path: Path)` instead of `path.toFile.isFile()`

Reply:
Changed to `!Files.isDirectory(path)` since it shows better the intent, and moved to `onChange`